fe/fe-core/src/main/java/com/starrocks/planner/SplitCastDataSink.java

```diff
@@ -0,0 +1,106 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.starrocks.planner;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.starrocks.analysis.Expr;
+import com.starrocks.thrift.TDataSink;
+import com.starrocks.thrift.TDataSinkType;
+import com.starrocks.thrift.TDataStreamSink;
+import com.starrocks.thrift.TExplainLevel;
+import com.starrocks.thrift.TExpr;
+import com.starrocks.thrift.TPlanFragmentDestination;
+import com.starrocks.thrift.TSplitDataStreamSink;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class SplitCastDataSink extends DataSink {
+    // SplitCastDataSink will have multiple dest fragment
+    private final List<DataStreamSink> dataStreamSinks = Lists.newArrayList();
+    // every DataStreamSink can have multiple destinations
+    private final List<List<TPlanFragmentDestination>> destinations = Lists.newArrayList();
+    // if splitExprs[i] is true for [0,i], then data will be sent to dest fragment i
+    private final List<Expr> splitExprs = Lists.newArrayList();
```

Comment:
suggest to adopt style like Collectors.groupBy(),  so we can cook a splitExpr from this splitExprs, the splitExpr always map input to 0..n-1 that correspond n partitions. for an example:

we can make a case-when expr from this splitExprs.
```
case 
when splitExprs[0]  then 0
when splitExprs[1] then 1
...
else n
end
```

Reply:
> suggest to adopt style like Collectors.groupBy(), so we can cook a splitExpr from this splitExprs, the splitExpr always map input to 0..n-1 that correspond n partitions. for an example:
> 
> we can make a case-when expr from this splitExprs.
> 
> ```
> case 
> when splitExprs[0]  then 0
> when splitExprs[1] then 1
> ...
> else n
> end
> ```
right now actually it's want you want. one chunk will first be evaluted by splitExprs[length - 1], if one row's result is true then it will be sent to dataStreamSinks[length - 1], if not, it will be  evaluted by splitExprs[length - 2] etc.  so every splitExprs[i] will cut cur_chunk into two chunk, one is sent to dataStreamSinks[i], another  is fed to splitExprs[i-1].